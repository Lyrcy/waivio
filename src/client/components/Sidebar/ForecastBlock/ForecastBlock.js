import React from 'react';
import _, {isEmpty} from 'lodash';
import PropTypes from 'prop-types';
import {Icon} from 'antd';
import {injectIntl} from 'react-intl';
import './ForecastBlock.less';
import ForecastItem from '../ForecastItem';

@injectIntl
class ForecastBlock extends React.Component {
  static propTypes = {
    forecastsByObject: PropTypes.arrayOf(PropTypes.shape({})),
    forecastsByUser: PropTypes.arrayOf(PropTypes.shape({})),
    getActiveForecasts: PropTypes.func.isRequired,
    intl: PropTypes.shape(),
    object: PropTypes.shape(),
  };

  static defaultProps = {
    getActiveForecasts: () => {
    },
    intl: {},
    object: {},
    forecastsByObject: [],
    forecastsByUser: [],
  };

  static forecastBlock(intl, forecasts) {
    return (
      <div className="forecasts-block">
        <h4 className="forecasts-block__header">
          <Icon type="rise" className="forecasts-block__header-icon"/>
          <span>
            {intl.formatMessage({
              id: 'forecast.currentForecast',
              defaultMessage: 'Current forecasts',
            })}
          </span>
        </h4>
        <div className="forecasts-block__content">
          {forecasts.slice(0, 5).map(forecast => (
            <ForecastItem
              key={forecast.id}
              permlink={forecast.id}
              quoteSecurity={forecast.security}
              recommend={forecast.recommend}
              forecast={forecast.forecast}
              postPrice={forecast.postPrice}
              dateTimeCreate={forecast.created_at}
            />
          ))}
        </div>
      </div>
    );
  }

  componentDidMount() {
    this.props.getActiveForecasts();
  }

  render() {
    const {forecastsByObject, forecastsByUser, intl, object} = this.props;
    if (!isEmpty(object)) {
      const chartId = _.find(object.fields, ['name', 'chartid']);
      console.log('CHART', chartId);
      return chartId && forecastsByObject && forecastsByObject.length
        ? ForecastBlock.forecastBlock(intl, forecastsByObject)
        : null;
    }
    return forecastsByUser && forecastsByUser.length
      ? ForecastBlock.forecastBlock(intl, forecastsByUser)
      : null;
  }
}

export default ForecastBlock;
