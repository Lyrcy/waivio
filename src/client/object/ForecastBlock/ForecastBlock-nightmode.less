@import (reference) '../../styles/custom.less';

.nightmode {
  .forecasts-block {
    border: 1px solid;
    border-color: @normal-grey !important;
    border-radius: 4px;
    background: @grey-nero;
    left: auto;
    top: auto;

    &__content {
      padding: 0 10px;
    }

    &__header {
      justify-content: flex-start;
      color: @blue-baliHai;
      font-size: 16px;
      font-weight: 500;
      display: flex;
      align-items: center;
      position: relative;
      background: @grey-nero;
      height: 45px;
      padding: 0 12px;
      border-bottom: 1px solid @normal-grey;
    }
  }
}
