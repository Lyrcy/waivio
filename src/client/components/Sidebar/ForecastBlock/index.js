import { connect } from 'react-redux';
import React from 'react';
import ForecastBlock from './ForecastBlock';
import { getActiveForecasts } from '../../../../investarena/redux/actions/forecastActions';
import { getForecastDataByUser } from '../../../reducers';

const ForecastBlockContainer = props => <ForecastBlock {...props} />;

const mapState = () => (state, ownProps) => ({
  forecasts: getForecastDataByUser(state, ownProps.username),
});

function mapDispatchToProps(dispatch) {
  return {
    getActiveForecasts: () => dispatch(getActiveForecasts()),
  };
}

export default connect(
  mapState,
  mapDispatchToProps,
)(ForecastBlockContainer);
