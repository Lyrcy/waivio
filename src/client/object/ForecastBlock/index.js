import { connect } from 'react-redux';
import React from 'react';
import ForecastBlock from './ForecastBlock';
import { getActiveForecasts } from '../../../investarena/redux/actions/forecastActions';
import { getForecastDataByQuote } from '../../reducers';

const ForecastBlockContainer = props => <ForecastBlock {...props} />;

const mapState = () => state => ({
  forecasts: getForecastDataByQuote(state, state.object.wobject.default_name),
  object: state.object.wobject,
});

function mapDispatchToProps(dispatch) {
  return {
    getActiveForecasts: () => dispatch(getActiveForecasts()),
  };
}

export default connect(
  mapState,
  mapDispatchToProps,
)(ForecastBlockContainer);
