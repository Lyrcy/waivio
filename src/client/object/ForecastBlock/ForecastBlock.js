import React from 'react';
import _ from 'lodash';
import PropTypes from 'prop-types';
import { Icon } from 'antd';
import { injectIntl } from 'react-intl';
import './ForecastBlock.less';
import './ForecastBlock-nightmode.less';
import ForecastItem from '../ForecastItem/index';

@injectIntl
class ForecastBlock extends React.Component {
  static propTypes = {
    getActiveForecasts: PropTypes.func,
    forecasts: PropTypes.shape(),
    intl: PropTypes.shape(),
  };

  static defaultProps = {
    getActiveForecasts: () => {},
    forecasts: {},
    intl: {},
  };

  componentDidMount() {
    this.props.getActiveForecasts();
  }

  render() {
    console.log('PROPS', this.props);
    const { forecasts, intl, object } = this.props;
    const chartId = _.find(object.fields, ['name', 'chartid']);
    return chartId && forecasts && forecasts.length ? (
      <div className="forecasts-block">
        <h4 className="forecasts-block__header">
          <Icon type="rise" className="forecasts-block__header-icon" />
          <span>
            {intl.formatMessage({
              id: 'forecast.currentForecast',
              defaultMessage: 'Current forecasts',
            })}
          </span>
        </h4>
        <div className="forecasts-block__content">
          {forecasts.slice(0, 5).map(forecast => (
            <ForecastItem
              key={forecast.id}
              quoteSecurity={forecast.security}
              recommend={forecast.recommend}
              forecast={forecast.forecast}
              postPrice={forecast.postPrice}
              dateTimeCreate={forecast.created_at}
            />
          ))}
        </div>
      </div>
    ) : null;
  }
}

export default ForecastBlock;
