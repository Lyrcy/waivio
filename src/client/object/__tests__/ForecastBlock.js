import configureMockStore from 'redux-mock-store';
import React from 'react';
import thunk from 'redux-thunk';
import ForecastBlock from '../ForecastBlock/ForecastBlock';
import { shallowWithStore, mountWithStore } from './shallowWrapper';

const mockStore = configureMockStore([thunk]);
const store = mockStore({
  quotes: {
    security: 'AUDCAD',
    bidPrice: '155.570',
    askPrice: '161.830',
    dailyChange: 3.7687,
    timestamp: 1562583219562,
    isSession: true,
    state: 'down',
  },
  quotesSettings: {
    ID: 1,
    market: 'Currency',
    name: 'AUD/CAD',
    priceRounding: 100000,
    leverage: 200,
    defaultQuantity: 500000000000,
    maximumQuantity: 40000000000000,
    minimumQuantity: 100000000000,
    quantityIncrement: 100000000000,
    tickSize: 100,
    popular: false,
    baseCurrency: 'AUD',
    termCurrency: 'CAD',
    keyName: 'AUDCAD',
  },
});

describe('<ForecastBlock />', () => {
  it('renders component without crashing', () => {
    const props = {
      username: 'z1wo5',
      forecasts: Array(5),
      getActiveForecasts: () => {},
    };
    const wrapper = shallowWithStore(<ForecastBlock {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders component without crashing whith props', () => {
    const props = {
      username: 'z1wo5',
      forecasts: [
        {
          author: 'z1wo5',
          created_at: '2019-07-04T11:02:36.000Z',
          forecast: '2019-07-04T15:02:36.000Z',
          id: 'z1wo5/aud-cad',
          permlink: 'aud-cad',
          postPrice: 0.91708,
          recommend: 'Sell',
          security: 'AUDACAD',
        },
      ],
      object: {
        author: 'sor31',
        createdAt: '2019-04-12T14:07:31.854Z',
        creator: 'wiv01',
        default_name: 'bitcoin',
        fields: [
          {
            author: 'vmn31',
            body: 'bitcoin',
            creator: 'wiv01',
            locale: 'en-US',
            name: 'chartid',
            permlink: 'wiv01-ccsvwze9lsl',
            weight: 1,
            _id: '5cb09ba6b81150400aa6bb92',
          },
        ],
      },
      getActiveForecasts: () => {},
    };
    const component = mountWithStore(<ForecastBlock {...props} />, store);
    const mainDiv = component.find('.st-forecast-wrap');
    console.log(component.html());
    expect(mainDiv).toHaveLength(1);
  });

  it('should render component in five forecasts case', () => {
    const props = {
      username: 'z1wo5',
      forecasts: Array(5),
      getActiveForecasts: () => {},
      object: {
        author: 'sor31',
        createdAt: '2019-04-12T14:07:31.854Z',
        creator: 'wiv01',
        default_name: 'bitcoin',
        fields: [
          {
            author: 'vmn31',
            body: 'bitcoin',
            creator: 'wiv01',
            locale: 'en-US',
            name: 'chartid',
            permlink: 'wiv01-ccsvwze9lsl',
            weight: 1,
            _id: '5cb09ba6b81150400aa6bb92',
          },
        ],
      },
    };
    const component = mountWithStore(<ForecastBlock {...props} />, store);
    const mainDiv = component.find('.forecasts-block');
    expect(mainDiv).toHaveLength(1);
  });

  it('should render component in more then five forecasts case', () => {
    const props = {
      username: 'z1wo5',
      forecasts: Array(6),
      getActiveForecasts: () => {},
      object: {
        author: 'sor31',
        createdAt: '2019-04-12T14:07:31.854Z',
        creator: 'wiv01',
        default_name: 'bitcoin',
        fields: [
          {
            author: 'vmn31',
            body: 'bitcoin',
            creator: 'wiv01',
            locale: 'en-US',
            name: 'chartid',
            permlink: 'wiv01-ccsvwze9lsl',
            weight: 1,
            _id: '5cb09ba6b81150400aa6bb92',
          },
        ],
      },
    };
    const component = mountWithStore(<ForecastBlock {...props} />, store);
    const mainDiv = component.find('.forecasts-block');
    console.log(component.html());
    expect(mainDiv).toHaveLength(1);
  });

  it('should not render component if forecasts is empty', () => {
    const props = {
      username: 'z1wo5',
      forecasts: [],
      getActiveForecasts: () => {},
      object: {
        author: 'sor31',
        createdAt: '2019-04-12T14:07:31.854Z',
        creator: 'wiv01',
        default_name: 'bitcoin',
        fields: [
          {
            author: 'vmn31',
            body: 'bitcoin',
            creator: 'wiv01',
            locale: 'en-US',
            name: 'chartid',
            permlink: 'wiv01-ccsvwze9lsl',
            weight: 1,
            _id: '5cb09ba6b81150400aa6bb92',
          },
        ],
      },
    };
    const component = mountWithStore(<ForecastBlock {...props} />, store);
    const mainDiv = component.find('.forecasts-block');
    expect(mainDiv).toHaveLength(0);
  });
});
