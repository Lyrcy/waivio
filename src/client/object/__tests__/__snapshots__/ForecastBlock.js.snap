// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ForecastBlock /> renders component without crashing 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    store={undefined}
  >
    <IntlProvider
      locale="en"
    >
      <MemoryRouter>
        <InjectIntl(ForecastBlock)
          forecasts={
            Array [
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
          getActiveForecasts={[Function]}
          username="z1wo5"
        />
      </MemoryRouter>
    </IntlProvider>
  </Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <MemoryRouter>
        <InjectIntl(ForecastBlock)
          forecasts={
            Array [
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
          getActiveForecasts={[Function]}
          username="z1wo5"
        />
      </MemoryRouter>,
      "locale": "en",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <InjectIntl(ForecastBlock)
          forecasts={
            Array [
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
            ]
          }
          getActiveForecasts={[Function]}
          username="z1wo5"
        />,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "forecasts": Array [
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ],
          "getActiveForecasts": [Function],
          "username": "z1wo5",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <MemoryRouter>
          <InjectIntl(ForecastBlock)
            forecasts={
              Array [
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
              ]
            }
            getActiveForecasts={[Function]}
            username="z1wo5"
          />
        </MemoryRouter>,
        "locale": "en",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <InjectIntl(ForecastBlock)
            forecasts={
              Array [
                undefined,
                undefined,
                undefined,
                undefined,
                undefined,
              ]
            }
            getActiveForecasts={[Function]}
            username="z1wo5"
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "forecasts": Array [
              undefined,
              undefined,
              undefined,
              undefined,
              undefined,
            ],
            "getActiveForecasts": [Function],
            "username": "z1wo5",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
